src/util/DRRScheduler.cpp

```diff
@@ -0,0 +1,155 @@
+// Copyright 2020 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "util/DRRScheduler.h"
+#include "medida/timer.h"
+#include <queue>
+
+namespace stellar
+{
+using usecs = std::chrono::microseconds;
+
+class DRRQueue
+{
+    // Each queue retains DURATION_MEMORY worth of action execution-durations,
+    // and estimates the next action durtion as the average over this memory.
+    static constexpr const size_t DURATION_MEMORY = 8;
```

Comment:
Actually even that approach won't necessarily work either:
if 99% of the messages are 1ms but 1% are ~1 second, we can't say that processing anything SCP related will take 1 second otherwise we'll fall out of sync.

I had a related question in the other PR:
why do we want to give priority to IO at the expense of consensus messages/closing ledgers?
In fact, I don't think it makes sense to throttle SCP traffic as much as what is proposed here (I was proposing more a of "real time" priority for SCP related traffic in the other PR).

Reply:
We could also try to refactor the SCP code to post externalization (this is tricky) but we'll need a way to not throttle that new queue as well, so it may not be worth trying (also, I am not 100% sure this would solve the problem: SCP messages are slower to process in general, so the SCP queues would fall behind regardless).